flink-ml-iteration/flink-ml-iteration-common/src/main/java/org/apache/flink/iteration/typeinfo/IterationRecordSerializer.java

```diff
@@ -106,7 +106,7 @@ public int getLength() {
     @Override
     public void serialize(IterationRecord<T> record, DataOutputView target) throws IOException {
         target.writeByte((byte) record.getType().ordinal());
-        serializerNumber(record.getEpoch(), target);
+        target.writeInt(record.getEpoch());
```

Comment:
The original `serializerNumber` cannot serialize `Integer.MAX_VALUE + 1`, which is the epoch of the feedback record in the last iteration. In the previous implementation, the feedback records are just passed in memory and don't need to serialize/deserialize, so the bug doesn't appear.

I also did a benchmark to serialize 1 billion integers with `serializerNumber` and `DataOutputSerializer#writeInt`, their performances are quite close, so I think we can remove the current `serializerNumber` implementation.

Reply:
Regarding "the original serializerNumber cannot serialize Integer.MAX_VALUE + 1", can you explain why it is necessary to pass `Integer.MAX_VALUE + 1` as the serializerNumber? Is it possible to change the caller side code to avoid that?

I agree the 1-byte overhead is likely not a big deal for common cases where iteration is needed. On the other hand, if it is possible to change the caller side code as appropriate without any downside, then that approach is better than increasing the epoch overhead from 1 byte to 4 bytes.

Reply:
Currently, the `Integer.MAX_VALUE + 1` is passed because the termination check is in the `HeadOperator` . When the `TailOperator` processed the last epoch_watermark, it needs to increment the epoch which is already `Integer.MAX_VALUE`, and send the increased epoch to the `HeadOperator`. Then the `TerminatingHeadOperatorRecordProcessor#processFeedbackElement` would check if the epoch is equal to `Integer.MAX_VALUE + 1` to determine whether to terminate the iteration.

Of course, we can avoid it. Maybe simply change the last epoch_watermark number to `Integer.MAX_VALUE - 1`, but I don't think we need to make the code less readable for this.

Reply:
How about we create something like `public final END_EPOCH_WATERMARK = Integer.MAX_VALUE - 1 ` and use this variable in e.g. `TerminatingHeadOperatorRecordProcessor`? Then the code can be readable without increasing metadata overhead.